### PR TITLE
Documenting text functions and improving documentation outline

### LINKED ISSUE #360: Documenting text functions and improving documentation outline
In the context recent changes to various text functions (most recently #343 / #350) it has come up that there is no in-depth explanation of those functions and their differences (let alone the new positioning options) other than in the docstrings. Since I've modifying some of that functionality already, I probably should also add such a page, though in a seperate PR to keep things manageable.
In addition, while looking through the docs, I found that the TOC seems to have grown a lot over time without anyone adding much structure to it, which makes it quite hard to find any specific information. I tried to play around with it a bit and this is the most reasonable structure I could come up with:

* Tutorial
    * English
    * Français
    * Deutsch
    * ...
* Page Layout
    * Page Format and Orientation
    * Margins
    * Templates
    * Tables
* Text Content
    * Adding Text (new)
    * Line Breaks
    * Page Breaks
    * Text Styling
    * Unicode
    * Emojis, Symbols & Dingbats
* Graphics Content
    * Images
    * Shapes
    * Barcodes
    * Drawing
    * SVG
    * Charts & Graphs (formerly "Maths")
* PDF Features
    * Links
    * Metadata
    * Annotations
    * Presentation Mode
    * Document Outline & Table of Contents
* Processing Web Content
    * HTML
    * Web2Py
* Reading and Postprocessing
    * pdfrw
    * Borb
* Development
    * Development Guidelines
    * Logging
* API
* FAQ

Any objections when I modify it that way, or any better ideas? There are a few items that could go in several places.
Do I understand this right, that the documentation outline is defined in mkdocs.yml? Is there an easy way to test its processing locally?

### PR DESCRIPTION
**Pull request by gmischler**
e.g. Fixes #360 

**Checklist**:

- [na] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.
- [na] A unit test is covering the code added / modified by this PR
- [x] This PR is ready to be merged 
- [na] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder
- [x] A mention of the change is present in `CHANGELOG.md`
- [x] The PR description or comment contains a picture of a cute animal (not mandatory but encouraged :pig:)

As the title says and explained in the linked issue.

Note that the new `Text.md` already describes the modified API from #350. If that should be too confusing, we can postpone that addition until everybody has made up their minds about the best deprecation strategy.

### CODE REVIEW

**Comment by Lucas-C** [comment]
The Hindi tutorial does not get rendered well in the PDF manual:
2022-03-19 10_15_35-Window
2022-03-19 10_18_32-Window

File: docs/Text.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
+# Adding Text
+
+There are several ways in fpdf to add text to a PDF document, each of which comes with its own special features and its own set of advantages and disadvantages. You will need to pick the right one for your specific task.
+
+| method | lines | markdown | HTML | positioning | details |
```
Comment (Lucas-C) [inline docs/Text.md:5]: Great idea to provide this comparison table! Good job

"markdown" -> "markdown support" ?

Also, the meaning ot the "positioning" column doesn't seem very clear to me... "control text alignement" instead maybe?

File: docs/Text.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+| -- | :--: | :--: | :--: | :--: | -- |
+| [`.text()`](#text)  | one | no | no | fixed | Inserts a single-line text string with a precise location on the base line of the font.|
+| [`.cell()`](#cell)  | one | yes | no | yes | Inserts a single-line text string within the boundaries of a given box, optionally with background and border. |
+| [`.multi_cell()`](#multi_cell) | several | yes | no | yes | Inserts a multi-line text string within the boundaries of a given box, optionally with background and border. |
+| [`.write()`](#write) | several | no | no | auto | Inserts a multi-line text string within the boundaries of the page margins, starting at the current x/y location (typically the end of the last inserted text). |
+| [`.write_html()`](#write_html) | several | no | yes | auto | From [html.py](HTML.html). An extension to `.write()`, with additional parsing of basic HTML tags.
+
+## Typographical Limitations
+
+There are a few advanced typesetting features that fpdf doesn't currently support.
```
Comment (Lucas-C) [inline docs/Text.md:15]: Could you please add a newline before lists? Otherwise it won't get rendered properly in HTML / PDF

File: docs/Text.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+* [`.set_draw_color()`](fpdf/fpdf.html#fpdf.fpdf.FPDF.set_draw_color) - for cell borders
+* [`.set_fill_color()`](fpdf/fpdf.html#fpdf.fpdf.FPDF.set_fill_color) - for the background
+
+In addition, some of the methods can optionally use [markdown](TextStyling.html#markdowntrue) or [HTML](HTML.html) markup in the supplied text in order to change the font style (bold/italic/underline) of parts of the output.
+
+## Change in current position
+`.cell()` and `.multi_cell()` let you specify where the current position (`.x`/`.y`) should go after the call. This is handled by the parameters `new_x` and `new_y`. Their values are one out of the Enums `.XPos` and `.YPos` respectively, which offer the following options:
+
+#### .XPos
+* XPos.LEFT    - left end of the cell
```
Comment (Lucas-C) [inline docs/Text.md:33]: Could you please add backquotes or double asterisks around those constant?

Also, wouldn't it be better to just point to the generated documentation in https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.XPos ? This way the fields descriptions will stay up-to-date if the code changes, no risk for this doc paragraph to become outdated.

File: docs/Text.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+
+## .text()
+Prints a single-line character string. In contrast to the other text methods,
+the position is given explicitly, and not taken from `.x`/`.y`. The origin is
+on the left of the first character, on the baseline. This method allows placing
+a string with typographical precision on the page, but it is usually easier to
+use the `.cell()`, `.multi_cell()` or `.write()` methods.
+
+#### text(x, y, txt='')
```
Comment (Lucas-C) [inline docs/Text.md:57]: I'd prefer to remove this paragraph and simply points to https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.text

File: docs/Text.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+#### .YPos
+* YPos.TOP     - top of the first line
+* YPos.LAST    - top of the last line (same as TOP for single-line text)
+* YPos.NEXT    - top of next line (bottom of current text)
+* YPos.TMARGIN - top page margin (start of printable area)
+* YPos.BMARGIN - bottom page margin (end of printable area)
+
+
+## .text()
+Prints a single-line character string. In contrast to the other text methods,
```
Comment (Lucas-C) [inline docs/Text.md:51]: 👍

File: docs/Text.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+* Automatic ligatures - Some writing systems (eg. most hindic scripts such as Devaganari, Hangul) frequently combine a number of written characters into a single glyph. This would require advanced font analysis capabilities, which aren't currently implemented.
+* Contextual forms - In some writing systems (eg. Arabic, Hebrew, Mongolian, etc.), characters may take a different shape, depending on whether they appear at the beginning, in the middle, or at the end of a word, or isolated. Fpdf will always use the same standard shape in those cases.
+* Vertical writing - Some writing systems are meant to be written vertically. Doing so is not directly supported. In cases where this just means to stack characters on top of each other (eg. Chinese, Japanese, etc.), client software can implement this by placing each character individuall at the correct location. In cases where the characters are connected with each other (eg. Mongolian), this may be more difficult, if possible at all.
+
+## Text Formatting
+For all text insertion methods, the relevant font related properties (eg. font/style and foreground/background color) must be set before invoking them. This includes using:
```
Comment (Lucas-C) [inline docs/Text.md:21]: 👍 
+ missing newline

File: docs/Text.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+position. The text can be aligned or centered. After the call, the current
+position moves to the selected `new_x`/`new_y` position. It is possible to put a link on the text.
+If `markdown=True`, then minimal [markdown](TextStyling.html#markdowntrue)
+styling is enabled, to render parts of the text in bold, italics, and/or
+underlined.
+
+If automatic page breaking is enabled and the cell goes beyond the limit, a
+page break is performed before outputting.
+
+#### cell(w=None, h=None, txt='', border=0, new_x=XPos.RIGHT, new_y=YPos.TOP, align='', fill=False, link='', center=False, markdown=False)
```
Comment (Lucas-C) [inline docs/Text.md:75]: Ditto: https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.cell

File: docs/Text.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+use the `.cell()`, `.multi_cell()` or `.write()` methods.
+
+#### text(x, y, txt='')
+* __x__ (float) - The horizontal component origin position
+* __y__ (float) - The vertical component origin position
+* __txt__ (str) - the text string to print
+
+
+## .cell()
+Prints a cell (rectangular area) with optional borders, background color and
```
Comment (Lucas-C) [inline docs/Text.md:64]: 👍

File: docs/Text.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+the background painted.
+
+Using `new_x=XPos.RIGHT, new_y=XPos.TOP, maximum height=pdf.font_size` can be
+useful to build tables with multiline text in cells.
+
+In normal operation, returns a boolean indicating if page break was triggered.
+When `split_only == True`, returns `txt` split into lines in an array (with any markdown markup removed).
+
+
+#### multi_cell(w, h=None, txt="", border=0, align="J", fill=False, split_only=False, link="", new_x=XPos.RIGHT, new_y=YPos.NEXT, max_line_height=None, markdown=False, print_sh=False) -> Bool or List
```
Comment (Lucas-C) [inline docs/Text.md:116]: Ditto: https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.multi_cell

File: docs/Text.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+Prints multi-line text between the page margins, starting from the current position.
+When the right margin is reached, a line break occurs at the most recent
+space or soft-hyphen character, and text continues from the left margin.
+A manual break happens any time the \\n character is met,
+Upon method exit, the current position is left near the end of the text, ready for the next call to continue without a gap, potentially with a different font or size set. Returns a boolean indicating if page break was triggered.
+
+The primary purpose of this method is to print continuously wrapping text, where different parts may be rendered in different fonts or font sizes. This contrasts eg. with `.multi_cell()`, where a change in font family or size can only become effective on a new line.
+
+
+#### write(h: float = None, txt: str = "", link: str = "", print_sh: bool = False) -> bool
```
Comment (Lucas-C) [inline docs/Text.md:160]: Just link to https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.write

File: docs/Text.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+* __print_sh__ (bool): Treat a soft-hyphen (\\u00ad) as a normal printable
+	character, instead of a line breaking opportunity.
+
+
+## .write_html()
+This method can be accessed by using the class `HTMLMixin` from "[html.py](HTML.html)". It is very similar to `.write()`, but accepts basic HTML formatted text as input. See [html.py](HTML.html) for more details and the supported HTML tags.
+
+Note that when using data from actual web pages, the result may not look exactly as expected, because `.write_html()` prints all whitespace unchanged as it finds them, while webbrowsers rather collapse each run of consequitive whitespace into a single space character.
+
+#### write_html(text, image_map=None, li_tag_indent=5, table_line_separators=False, ul_bullet_char=BULLET_WIN1252, heading_sizes=None)
```
Comment (Lucas-C) [inline docs/Text.md:175]: Just link to https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.write_html

File: mkdocs.yml [hunk 12]
```diff
@@ -51,33 +51,49 @@ plugins:
...
-- 'SVG':                            'SVG.md'
-- 'Barcodes':                       'Barcodes.md'
-- 'Metadata':                       'Metadata.md'
-- 'Logging':                        'Logging.md'
-- 'Templates':                      'Templates.md'
-- 'Unicode':                        'Unicode.md'
-- 'Emojis, Symbols & Dingbats':     'EmojisSymbolsDingbats.md'
-- 'borb':                           'borb.md'
-- 'Web2Py':                         'Web2Py.md'
+- 'Tutorial':
```
Comment (Lucas-C) [inline mkdocs.yml:54]: ❤️

**Review by Lucas-C — APPROVED** [review]

File: docs/Text.md (hunk 1 above)
Comment (gmischler) [reply docs/Text.md:5]: "markdown support" and "HTML support" will make it more explicit.
positioning is not related to alignment. "Accepts new end position" explains it better.

File: docs/Text.md (hunk 2 above)
Comment (gmischler) [reply docs/Text.md:15]: Will do.

File: docs/Text.md (hunk 3 above)
Comment (gmischler) [reply docs/Text.md:33]: Highlighting them is a good idea.

Not sure about replacing them with links here.
(Yes, all documentation needs updating on API changes.)

File: docs/Text.md (hunk 4 above)
Comment (gmischler) [reply docs/Text.md:57]: Ah yes, with all of them, I can see your point now.

(Currently travelling with a wonky laptop keyboard, so not looking forward to any non-trivial changes 🙄 )

**Comment by Lucas-C** [comment]
Once this is merged I'll add some code snippet from https://github.com/PyFPDF/fpdf2/issues/116#issuecomment-1081780278 to use `fpdf2` in a Django app.

**Comment by gmischler** [comment]
> The Hindi tutorial does not get rendered well in the PDF manual:

Has it ever rendered correctly?
It looks like the markdown-PDF converter needs to be told which font to use.

**Comment by Lucas-C** [comment]
> Has it ever rendered correctly?

I think this tutorial simply wasn't included in the PDF manual beforehand.

> It looks like the markdown-PDF converter needs to be told which font to use.

OK. This could be fixed later on.

Could you just rebase on `master` in order to fix the conflick in `mkdocs.yml` please?
Then I'll merge this.

**Comment by Lucas-C** [comment]
Actually I resolved the conflict.
Merging now.

**Pull request merged**